Lighthouse ci

thatbudakguy opened the pull request:
this PR implements lighthouse in CI using the lighthouse-ci github action. see `lighthouserc.js` for the configuration of lighthouse itself; details on options can be found [on the lighthouse repo](https://github.com/GoogleChrome/lighthouse-ci/blob/master/docs/configuration.md).

rlskoeser commented on lighthouserc.js:
```diff
@@ -0,0 +1,42 @@
+    },
+    assert: {
+      assertMatrix: [
+        // homepage/issue detail incorrectly error on contrast of background
+        // text with role=presentation
```
I don't quite understand this comment — am I understanding it's wrongly detecting an error and you're configuring it to ignore? If so, might be good to add a little more detail about what the problem is.

Will we be able to turn this off at some point? Or if not, is there a chance we will miss other problems because of this configuration?

thatbudakguy replied:
I should expand this, thanks. Basically the transparent background text that appears on the homepage and issue detail page that's generated from the essay previews was generating contrast warnings, but that text isn't really meant to be read (it's decorative), so I gave it `role=presentation`. Lighthouse (so I guess aXe) still flags it...I thought about filing an issue, but for now the best we can do is ignore contrast errors for those pages specifically (I don't think you can get more granular than that in Lighthouse).

rlskoeser commented on lighthouserc.js:
```diff
@@ -0,0 +1,42 @@
+          // global settings for all pages
+          "matchingUrlPattern": ".*",            
+          assertions: {
+              // warn if overall performance score is lower than 85
+              "categories:performance": ["warn", { minScore: 0.85 }],
+              // error if overall accessibility score is lower than 100
+              "categories:accessibility": ["error", { minScore: 1 }],
```
looks neat to be able to configure it this way

rlskoeser commented on lighthouserc.js:
```diff
@@ -0,0 +1,42 @@
+module.exports = {
```
Seems like it would be helpful to add some comments here. Might be helpful to link to the documentation for the structure of this file and configuration options.

rlskoeser reviewed (approved):
Looks good. My questions are largely around comments / documentation.

The checks generated by the action look pretty helpful and usable enough.

thatbudakguy commented on .github/workflows/lighthouse-ci.yml:
```diff
@@ -0,0 +1,46 @@
+name: Lighthouse CI
+on: [push]
+env:
+    HUGO_VERSION: 0.67.1
+    NODE_VERSION: 10
```
You can use a combination of the [`engines` property in `package.json`](https://docs.npmjs.com/files/package.json#engines) and an `.npmrc` to select and enforce using a particular node version, so we could try that.

For Hugo, I'm not sure...is there a convention? It looks like [Netlify uses an environment variable](https://community.netlify.com/t/how-do-i-change-the-version-of-hugo-used-to-build-my-site/17883). We could try creating a `.env` file that gets `source`d, or I think there are a bunch of fancier solutions for automatically setting up an environment too.

rlskoeser replied:
An env file that gets sourced seems reasonable, but let's not put effort into anything fancy  — minimum solution could just be to document it so we remember to change both. I don't have a sense yet of how often we'll want or need to change it

thatbudakguy replied:
Makes sense. We have some language in the README now about checking the badge to make sure you're using the right version of hugo, so perhaps it could go with that.

What about if we create an Asana template task for publishing issues (like we do for software releases), and that task includes updating Hugo to latest version and then committing the change here (and on the README badge) if tests pass?

thatbudakguy replied:
Created an Asana template task with instructions for updating the Hugo version we test against/build with as part of issue releases.

thatbudakguy commented on .github/workflows/lighthouse-ci.yml:
```diff
@@ -0,0 +1,46 @@
+env:
+    HUGO_VERSION: 0.67.1
+    NODE_VERSION: 10
+    LHCI_GITHUB_APP_TOKEN: ${{ secrets.LHCI_GITHUB_APP_TOKEN }}
```
It's required to enable the github status checks that appear on PRs and commits; see [the GitHub app](https://github.com/apps/lighthouse-ci).

thatbudakguy commented:
@rlskoeser ok, I've made some changes and I think this is ready for another look. the big one is switching from the `assertMatrix` strategy (which has confusing options and doesn't appear well-supported) to just using a lighthouse preset and controlling individual checks. i've switched the two audits that cause unavoidable errors (contrast and font-swapping) to `warn` so that we don't lose any information if new violations occur, and added notes to the config file about them. the other set of rules is for recommended performance audits that cause errors, which we can enable later on when/if we decide to address them.

for the github actions workflows, I harmonized the deployment and ci workflows, and added some docs/comments to both.

rlskoeser reviewed (approved):
Updates look good!
